lang/functions_test.go

```diff
@@ -1039,32 +1052,89 @@ func TestFunctions(t *testing.T) {
 		},
 	}
 
-	data := &dataForTests{} // no variables available; we only need literals here
-	scope := &Scope{
-		Data:    data,
-		BaseDir: "./testdata/functions-test", // for the functions that read from the filesystem
-	}
+	experimentalFuncs := map[string]experiments.Experiment{}
+	experimentalFuncs["defaults"] = experiments.ModuleVariableOptionalAttrs
 
-	// Check that there is at least one test case for each function, omitting
-	// those functions that do not return consistent values
-	allFunctions := scope.Functions()
+	t.Run("all functions are tested", func(t *testing.T) {
+		data := &dataForTests{} // no variables available; we only need literals here
+		scope := &Scope{
+			Data:    data,
+			BaseDir: "./testdata/functions-test", // for the functions that read from the filesystem
+		}
 
-	// TODO: we can test the impure functions partially by configuring the scope
-	// with PureOnly: true and then verify that they return unknown values of a
-	// suitable type.
-	for _, impureFunc := range impureFunctions {
-		delete(allFunctions, impureFunc)
-	}
-	for f, _ := range scope.Functions() {
-		if _, ok := tests[f]; !ok {
-			t.Errorf("Missing test for function %s\n", f)
+		// Check that there is at least one test case for each function, omitting
+		// those functions that do not return consistent values
+		allFunctions := scope.Functions()
+
+		// TODO: we can test the impure functions partially by configuring the scope
+		// with PureOnly: true and then verify that they return unknown values of a
+		// suitable type.
+		for _, impureFunc := range impureFunctions {
+			delete(allFunctions, impureFunc)
 		}
-	}
+		for f := range scope.Functions() {
+			if _, ok := tests[f]; !ok {
+				t.Errorf("Missing test for function %s\n", f)
+			}
+		}
+	})
 
 	for funcName, funcTests := range tests {
 		t.Run(funcName, func(t *testing.T) {
+
+			// prepareScope starts as a no-op, but if a function is marked as
+			// experimental in our experimentalFuncs table above then we'll
+			// reassign this to be a function that activates the appropriate
+			// experiment.
+			prepareScope := func(t *testing.T, scope *Scope) {}
+
+			if experiment, isExperimental := experimentalFuncs[funcName]; isExperimental {
```

Comment:
This definitely achieves the goal of testing both cases (experimental and non-experimental), but I think got stuck in the idea that this all has to happen in one test (which I don't believe to be the case). 

I found this a bit difficult to understand (multiple loops through the test cases) and might be to future readers as well, so I will share that but approve this as it does have test coverage, but I don't think "it all must be tested in one large test" is a constraint that needs to be prioritized in future work.

Reply:
This particular test has indeed grown a bit beyond its original purpose of "make sure every registered function is at least callable", but I'm not really sure how to separate out dealing with experiments because by definition they make it so that registered functions are _not_ callable, and so the test would at least need to know enough to skip them so another test could deal with them, and then we'd have _two_ tests that both have to be aware of which functions are experimental, which I don't think would really get the usual benefits of having separated tests (that is: them each being simple and testing one specific thing).

Indeed though, we can probably find another way to deal with this down the road.